In [26]:
from data_handling_for_heuristic import *
import random as rn
from datetime import datetime

# 주의: eng.train과 train.txt의 B, I가 일치되지 않는다. train.txt 파일을 사용하자.
#read_path = 'data/conll2003/eng.train' #
read_path = 'train.txt'
raw_list, label_list = load_conll2003(read_path)
len(raw_list)

14987

In [3]:
raw_list[14986]

'Swansea 1 Lincoln 2'

### Count the number of sents that only have MISC

In [ ]:
# cnt = 0
# for sent in label_list:
#     splited_sent = sent.split()
#     if 'B-MISC' in splited_sent:
#         if not 'B-PER' in splited_sent:
#             if not 'B-ORG' in splited_sent:
#                 if not 'B-LOC' in splited_sent:
#                     cnt += 1
#                     #print(sent)
# print(cnt)

# cnt = 0
# for sent in label_list:
#     splited_sent = sent.split()
#     tkn = False
#     for token in splited_sent:
#         if token != 'O': # 엔티티가 하나라도 있으면.
#             tkn = True
#     if tkn == False: # 엔티티가 하나라도 없었다면.
#         cnt += 1
# print(cnt)

1. delete: 1049, insert: 4853, move: 382, replace: 4917  = **TOTAL : 11201**
2. 거의 2배 (MISC빼고, 엔티티가 하나도 없는거 빼고) = **TOTAL : 14987 - 930 - 3855 = 10202**

# 1. 단어 base로 Copy&Paste

In [29]:
req_n_word = 1138554 # 1190772 - 52218

new_sent_raw_list = []
new_sent_label_list = []
new_sent_raw_list += raw_list
new_sent_label_list += label_list
sent_idx = 0
total_n_word = 0

# 원본 total word 개수: 204570
for sent in raw_list:
    total_n_word += len(sent.split())

while(True):

    new_sent_raw_list.append(raw_list[sent_idx])
    new_sent_label_list.append(label_list[sent_idx])    
    
    total_n_word += len(raw_list[sent_idx].split())
    sent_idx += 1
    
    if sent_idx == len(raw_list)-1:
        sent_idx = 0
    
    if total_n_word >= req_n_word:
        print('req_n_word:', req_n_word)
        print('total_n_word:', total_n_word)
        print('total_n_sent:', len(new_sent_raw_list))
        break


##################        
""" 저장하기 """
##################
w_filename = ''
w_filename += '(wordbase)'
w_filename += datetime.now().strftime('%Y-%m-%d')
w_filename += ('_' + str(total_n_word))
path_write = 'gen_data/copy/' + w_filename + '.txt'
#path_write = 'test.txt'

with open(path_write, 'w', encoding='UTF-8') as txt:
    for i, _ in enumerate(new_sent_raw_list):
        splited_sent = new_sent_raw_list[i].split()
        splited_label = new_sent_label_list[i].split()
        for j, token in enumerate(splited_sent):
            txt.write(splited_sent[j]+' '+'NNP'+' '+'B-NP'+' '+splited_label[j])
            txt.write('\n')
        txt.write('\n')

req_n_word: 1138554
total_n_word: 1138589
total_n_sent: 84060


# 2. 문장base로 Copy&Paste

In [7]:
new_sent_raw_list = []
new_sent_label_list = []


num_of_sent = 10202

for i in range(num_of_sent):
    gen_num = (rn.randrange(1,len(raw_list)))
    new_sent_raw_list.append(raw_list[gen_num])
    new_sent_label_list.append(label_list[gen_num])
    
#total_sent_raw_list = raw_list + new_sent_raw_list
#total_sent_label_list = label_list + new_sent_label_list

#len(total_sent_raw_list)

In [8]:
total_sent_raw_list = new_sent_raw_list #+ raw_list  
total_sent_label_list = new_sent_label_list #+ label_list 
len(total_sent_raw_list)

10202

## Store new sentence

In [9]:
path_write = 'warehouse_data_data/ran(10202).txt'
#path_write = 'test.txt'

with open(path_write, 'w', encoding='UTF-8') as txt:
    
    for i, _ in enumerate(total_sent_raw_list):
        splited_sent = total_sent_raw_list[i].split()
        splited_label = total_sent_label_list[i].split()
        for j, token in enumerate(splited_sent):
            txt.write(splited_sent[j]+' '+'NNP'+' '+'B-NP'+' '+splited_label[j])
            txt.write('\n')
        txt.write('\n')

In [30]:
a = ['a','','','b']
a.remove('')

In [31]:
a

['a', '', 'b']

In [33]:
x = [1,2,3,'','','',3,4]
list(filter(lambda a: a != '', x))

[1, 2, 3, 3, 4]